src/shared/cpu-set-util.c

```diff
@@ -287,3 +295,98 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
         s = (CPUSet) {};
         return 0;
 }
+
+void numa_policy_set_prefer_if_unset(NUMAPolicy *policy) {
+        assert(policy);
+
+       /* In case when NUMAMask= is set but policy is left unspecified we want to use "prefer" policy.
+          This function should be called after NUMAMask= was parsed. If the NUMAPolicy= is parsed
+          later it will override the default that we set here. */
+        if (policy->nodes.allocated != 0 && policy->type == -1)
+                policy->type = MPOL_PREFERRED;
```

Comment:
please don't actually update the field here. This means the information that we automatically picked the preferred setting is lost that way, and this matters for example if inside the same unit file you first set a mask, and later unset it again. the first setting will fix the policy to preferred, and that will stick even though setting and then unsetting a unit option really should undo everything.

My recommendation: never update the NUMAPolicy structure with the automatically determined value. Instead, define an accessor function:

 ```c
int numa_policy_get_type(NUMAPolicy *policy) {
        assert(policy);
        return policy->type < 0 ? (policy->nodes.set ? MPOL_PREFERRED : MPOL_DEFAULT) : policy->type;
}
```

And then, when ever you want to know the numa policy type just write `numa_policy_get_type(p)` instead of `p->type`

Reply:
We can't do that I think. Looking at `execute.c` where we apply the policy for new units,

```C
        if (mpol_is_valid(context->numa_policy.type)) {
                r = apply_numa_policy(&context->numa_policy);
                if (r == -EOPNOTSUPP)
                        log_unit_debug_errno(unit, SYNTHETIC_ERRNO(r), "NUMA support not available, ignoring.");
                else if (r < 0) {
                        *exit_status = EXIT_NUMA_POLICY;
                        return log_unit_error_errno(unit, r, "Failed to set NUMA memory policy: %m");
                }
        }
```

If I take your suggestion then we would violate the current semantics. Because we don't want to reset policy to default if none was set. We want to inherit policy from the manager.

Reply:
then maybe change the accessor to:

```c
return policy->type < 0 ? (policy->nodes.set ? MPOL_PREFERRED : -1) : policy->type;
```

i.e. propagate the -1 if no nodes are set, so that you then know it's not set.

Reply:
Yes, that should work